File: app/lib/deepseek.js
const DEFAULT_BASE_URL = "https://api.deepseek.com";

export function getDeepSeekConfig() {
  return {
    apiKey: process.env.DEEPSEEK_API_KEY,
    model: process.env.DEEPSEEK_MODEL,
    baseUrl: process.env.DEEPSEEK_BASE_URL || DEFAULT_BASE_URL
  };
}

export async function callDeepSeek(messages, options = {}) {
  const config = getDeepSeekConfig();
  const model = options.model || config.model;

  if (!config.apiKey) {
    throw new Error("DEEPSEEK_API_KEY is not configured");
  }
  if (!model) {
    throw new Error("DEEPSEEK_MODEL is not configured");
  }

  const response = await fetch(`${config.baseUrl.replace(/\/$/, "")}/chat/completions`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${config.apiKey}`,
      "Content-Type": "application/json"
    },
    body: JSON.stringify({
      model,
      messages,
      temperature: options.temperature ?? 0.2
    })
  });

  const text = await response.text();
  let data;
  try {
    data = JSON.parse(text);
  } catch {
    data = { raw: text };
  }

  if (!response.ok) {
    const message = data?.error?.message || data?.message || `DeepSeek API error ${response.status}`;
    const error = new Error(message);
    error.detail = data;
    error.status = response.status;
    throw error;
  }

  return {
    answer: data?.choices?.[0]?.message?.content || "",
    raw: data
  };
}

File: src/stock_agent/providers/deepseek.py
from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any, List, Optional

from ..core import ConversationMessage


@dataclass
class DeepSeekConfig:
    api_key: Optional[str] = None
    base_url: str = "https://api.deepseek.com"
    model: str = ""

    @classmethod
    def from_env(cls) -> "DeepSeekConfig":
        return cls(
            api_key=os.getenv("DEEPSEEK_API_KEY"),
            base_url=os.getenv("DEEPSEEK_BASE_URL", "https://api.deepseek.com"),
            model=os.getenv("DEEPSEEK_MODEL", ""),
        )


class DeepSeekProvider:
    def __init__(self, config: Optional[DeepSeekConfig] = None) -> None:
        self.config = config or DeepSeekConfig.from_env()

    def complete(self, messages: List[ConversationMessage], **kwargs: Any) -> str:
        if not self.config.api_key:
            raise RuntimeError(
                "DeepSeek API key is missing. Set DEEPSEEK_API_KEY before calling complete()."
            )
        if not self.config.model:
            raise RuntimeError(
                "DeepSeek model is missing. Set DEEPSEEK_MODEL before calling complete()."
            )

        payload = {
            "model": self.config.model,
            "messages": [
                {
                    "role": message.role.value,
                    "content": message.content,
                    **({"name": message.name} if message.name else {}),
                }
                for message in messages
            ],
        }

        request = urllib.request.Request(
            url=self.config.base_url.rstrip("/") + "/chat/completions",
            data=json.dumps(payload).encode("utf-8"),
            headers={
                "Authorization": f"Bearer {self.config.api_key}",
                "Content-Type": "application/json",
            },
            method="POST",
        )

        try:
            with urllib.request.urlopen(request, timeout=120) as response:
                data = json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            detail = exc.read().decode("utf-8", errors="replace")
            raise RuntimeError(f"DeepSeek API error {exc.code}: {detail}") from exc
        except urllib.error.URLError as exc:
            raise RuntimeError(f"DeepSeek API connection error: {exc.reason}") from exc

        choices = data.get("choices") or []
        if not choices:
            raise RuntimeError("DeepSeek API returned no choices.")

        message = choices[0].get("message") or {}
        content = message.get("content")
        if content is None:
            return ""
        return content
